File: Kotlin/rotate-list.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun rotateRight(head: ListNode?, k: Int): ListNode? {
        if (k == 0 || head == null || head.next == null) {
            return head
        }
        val tmp = mutableListOf<ListNode?>()
        var n = head
        while (n != null) {
            tmp.add(n)
            n = n.next
        }
        if (k % tmp.size == 0) {
            return head
        }
        val split = tmp.size - k % tmp.size
        tmp[split-1]!!.next = null
        tmp.last()!!.next = head        
        return  tmp[split]
    }
}

File: JavaScript/rotate-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var rotateRight = function (head, k) {
    // calculate total list size
    let size = 0
    let tail = null
    for (let node = head; node; node = node.next) {
        size++
        if (node.next == null) {
            tail = node
        }
    }
    // handle cycles...
    k %= size
    if (size < 2 || k === 0) {
        return head
    }
    const splitPoint = size - k
    let beforeSplit = null
    let splitNode = head
    for (let i = 0; i < splitPoint; i++) {
        beforeSplit = splitNode
        splitNode = splitNode.next
    }
    const newHead = splitNode
    tail.next = head
    beforeSplit.next = null
    return newHead
};